### PR TITLE
On Demand Cookbook Libraries RFC

### PR DESCRIPTION
**Pull request by jkeiser**
**Breaks Backwards Compatibility? (Workstation):** No
**Breaks Backwards Compatibility? (Cookbooks):** No
**User Benefit:** Med

### CODE REVIEW

**Comment by ranjib** [comment]
:+1:  a shorter an sweeter method name will be awesome :-)

**Comment by jkeiser** [comment]
@ranjib totally cool with other names than `load_libraries_on_demand`.  But not sure what :)  I'd like to pick something that is specific enough to say what it does as well as simple enough for someone who isn't familiar with Ruby load paths and requires to understand basically what's going on.

**Comment by coderanger** [comment]
`on_demand` implies some kind of autoload-style behavior, such as how Rails loads library code. That should be avoided at all costs as a system-level feature. It has led to tons of issues in Rails.

**Comment by jkeiser** [comment]
There is no lazy loading in this proposal, thankfully.

**Comment by coderanger** [comment]
Should probably rename it then :-) Or at least make it very clear that "on-demand" != "autoload".

**Comment by ranjib** [comment]
```ruby
add_library true
```
?

**Comment by jeremyolliver** [comment]
Documentation of this feature should probably point out that files intended for use in this manner with `require`, will need to be named to avoid collisions with other cookbook library files, and even trickier to avoid collisions with ruby stdlib, and gem names in the chef omnibus install correct?

**Comment by ranjib** [comment]
@jeremyolliver your concerns are right, but this risk portfolio is same as any other gem or library, so im unsure if we have to explicitly call this out

**Comment by coderanger** [comment]
Having mulled it, I would probably call this feature "eager loading". So `eager_load_libraries false` to disable it. That also nicely sets the stage to make it the default in Chef 13 (a man can dream).

**Comment by jonlives** [comment]
I like this idea, again though I'm not sure what the best name for it is - not being familiar with the rails context of the term I actually quite liked `load_libraries_on_demand`.

**Comment by jkeiser** [comment]
`eager_load_libraries` is fine with me, and fits with the idea that eagerly loading stuff is the unusual action (which IMO it is).

**Comment by coderanger** [comment]
:+1: Also as a corollary between this and the chef_version field being added, I think it shows we should reconsider how to handle unknown metadata on the server side, maybe make it more of a non-validating hash. That's out of scope for this RFC but will probably factor in to the implementation :-)

**Comment by thommay** [comment]
:+1:

File: new/on-demand-cookbook-libraries.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+```ruby
+# metadata.rb
+name "mylibrary"
+version "0.0.1"
+eager_load_libraries false
+```
+
+If `eager_load_libraries` is false, Chef Client would instead append it to the ruby load path.  This would occur in the phase where the `libraries` files of a cookbook otherwise be automatically required in alphabetical order.  All recipes, libraries, and other Ruby code running in the Chef Client can then use `require 'filename'`` and it will load `cookbooks/mycookbook/libraries/filename` if present.
+
+If a file named `default.rb` exists in `libraries`, it will be required automatically.  No other files will be automatically loaded if `eager_load_libraries` is false.
```
Comment (lamont-granquist) [inline new/on-demand-cookbook-libraries.md]: Not wild about the magic default.rb here.  I see where you want to bootstrap the library loading with a hook, but maybe this should go into the metadata parameter?

`eager_load_libraries [ 'default.rb' ]`

Allow the filenames to contain globs and then the user has control and its explicit with no magic.

Comment (jkeiser) [inline new/on-demand-cookbook-libraries.md]: Sure.  With eager_load_libraries, that actually makes even more sense.  I'll change it.

**Comment by lamont-granquist** [comment]
:+1:

**Comment by jonlives** [comment]
:+1:

**Comment by adamhjk** [comment]
I am the decider, and I approve this message. GoGo Gadget @chef/rfc-editors

**Comment by martinisoft** [comment]
:+1:

**Comment by jtimberman** [comment]
Per IRC, known tools that might have their own custom metadata loading that will need to be aware of this new metadata field:

* [stove](https://github.com/sethvargo/stove/blob/master/lib/stove/cookbook/metadata.rb)
* [ridley](https://github.com/reset/ridley/blob/master/lib/ridley/chef/cookbook/metadata.rb) (via berkshelf)

I'm not aware of other tools offhand that implement their own metadata parsing class. The librarian-chef project uses [Chef::Cookbook::Metadata](https://github.com/applicationsonline/librarian-chef/blob/603de858cef9e4e4a25f5013f28d43e218b942b7/lib/librarian/chef/manifest_reader.rb#L29).

**Pull request merged**